src/mo_frontend/stability.ml

```diff
@@ -29,39 +29,39 @@ let error_sub s tf1 tf2 =
         display_typ_expand tf1.typ
         display_typ_expand tf2.typ))
 
-let error_mut s tf1 tf2 =
-  Diag.add_msg s
-    (Diag.error_message Source.no_region "M0171" cat
-      (Format.asprintf "stable variable %s changes mutability from previous type%a\nto new type %a"
-         tf1.lab
-         display_typ_expand tf1.typ
-         display_typ_expand tf2.typ))
+(* Relaxed rules with enhanced orthogonal persistence for more flexible upgrades.
+   - Mutability of stable fields can be changed because they are never aliased.
+   - Stable fields can be dropped, however, with a warning of potential data loss. 
+     For this, we give up the transitivity property of upgrades.
 
+   Upgrade transitivity means that an upgrade from a program A to B and then from B to C 
+   should have the same effect as directly upgrading from A to C. If B discards a field 
+   and C re-adds it, this transitivity is no longer maintained. However, rigorous upgrade 
+   transitivity was also not guaranteed before, since B may contain initialization logic
+   or pre-/post-upgrade hooks that alter the stable data.
+*)
 let match_stab_sig tfs1 tfs2 : unit Diag.result =
   (* Assume that tfs1 and tfs2 are sorted. *)
   let res = Diag.with_message_store (fun s ->
-    (* Should we insist on monotonic preservation of fields, or relax? *)
     let rec go tfs1 tfs2 = match tfs1, tfs2 with
       | [], _ ->
         Some () (* no or additional fields ok *)
       | tf1 :: tfs1', [] ->
-        error_discard s tf1;
-        go tfs1' []
+        (* dropping fields is allowed, but with a warning *)
+        warning_discard s tf1;
```

Comment:
I guess this only warns about the first field, but not the remaining ones in tfs1'. That was fine for an error (just bail on the first error) but is less fine for a warning... 
Maybe replace by `List.iter (warning_discard s) tfs1`?

Reply:
Or just recurse down tfs1' [] as in the original code. That should issue all the warnings, just as it issued all the errors.

Reply:
Thank you for having found this issue. I re-added the recursion.